File: LeetcodeJavaSolutions/src/other_problems/checkIn10/pathSum.js
function TreeNode(val, left, right) {
    this.val = (val === undefined) ? 0 : val;
    this.left = (left === undefined) ? null : left;
    this.right = (right === undefined) ? null : right;
}

var pathSum = function(root, targetSum) {
    let helper = function(root, target, map, curSum) {
        if (!root) return 0;
        curSum += root.val;
        
        let ans = 0;
        ans += map.get(curSum - target) || 0;
        const curSumVal = map.get(curSum);
        const newVal = curSumVal ? curSumVal + 1 : 1;
        map.set(curSum, newVal);

        ans += helper(root.left, target, map, curSum);
        ans += helper(root.right, target, map, curSum);

        const oldVal = map.get(curSum) - 1;
        map.set(curSum, oldVal);

        return ans;
    }

    const map = new Map();
    map.set(0, 1);
    return helper(root, targetSum, map, 0);
}

pathSum([10,5,-3,3,2,null,11,3,-2,null,1], 8)

File: LeetcodeJavaSolutions/src/binary_tree/problem112/PathSum.java
package binary_tree.problem112;

import java.util.LinkedList;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode (int x) {
        val = x;
        left = null;
        right = null;
    }
}

class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) return false;
        LinkedList<TreeNode> node_stack = new LinkedList();
        LinkedList<Integer> sum_stack = new LinkedList();
        node_stack.add(root);
        sum_stack.add(targetSum - root.val);
        TreeNode node;
        int curr_sum;
        while (!node_stack.isEmpty()) {
            node = node_stack.pollLast();
            curr_sum = sum_stack.pollLast();
            if (node.right == null && node.left == null && curr_sum == 0) return true;
            if (node.right != null) {
                node_stack.add(node.right);
                sum_stack.add(curr_sum - node.right.val);
            }
            if (node.left != null) {
                node_stack.add(node.left);
                sum_stack.add(curr_sum - node.left.val);
            }
        }
        return false;
    }
}

public class PathSum {
    
}
